### ISSUE TITLE
Weird treatment of trailing HTML

### ISSUE THREAD

**Issue by c-f-h** [Python-Markdown/markdown#572]
The following input

```
<!DOCTYPE html>
<html><body>

some text here

</body></html>
```

when run through the CLI creates the output

```
<!DOCTYPE html>
<html><body>

<p>some text here</p>
<p></body></html></p>
```

Obviously, the `<p>` tags around the trailing HTML code are wrong. Weirdly, if the leading and trailing HTML code are a simple `<div>` tag, these superfluous `<p>` tags are not produced.

For reference, the Python CommonMark implementation does the right thing here.

**Comment by facelessuser** [comment]
The problem you are seeing is specifically because Python Markdown doesn't handle `body` or `html`. It's supposed to not convert content between block tags, but it doesn't account for `body` or `html`.

Usually people are running content through and inserting it in HTML (or basically wrapping `body` and `html` around it after conversion), not usually running a fully formed HTML document through the parser.  You'll notice Python Markdown never wraps content in `body` and `html` when it outputs content.  It really is a content parser...

With that said, maybe it should at least account for `html` and `body` so it doesn't surprise people.  If @waylan agrees, I wouldn't mind issuing a pull request, but if not, my best advice is don't feed your content through the parser wrapped in `html` and `body`.

**Comment by c-f-h** [comment]
But why does it work for the opening `html` and `body` tags, but not the closing ones?

Also I thought that Markdown is intended to leave any inline HTML alone, which is why I thought the above behavior is a bug.

**Comment by facelessuser** [comment]
I haven't looked at the code for specifics, but content **between** blocks should not get parsed.  In your example, the content and the closing tags are getting parsed.  This is telling me the html blocks `html` and `body` aren't getting handled like block tags.  Most likely they are getting handled like inline tags, but since you have empty new lines after the last opening tag and before the first closing tag, the opening and closing tags are in different paragraph blocks breaking the inline tag treatment.

Basically, `html` and `body` are probably being treated as inline tags, but are spanning across multiple paragraphs which breaks parsing.

**Comment by waylan** [comment]
@facelessuser is correct. The `html` and `body` tags are seen as inline tags. The reason the opening tags are not wrapped in a `p` element is because of the doctype. If, instead, you do this:

```html
<!DOCTYPE html>

<html><body>

some text here

</body></html>
```

Then you get the following output:

```html
<!DOCTYPE html>

<p><html><body></p>
<p>some text here</p>
<p></body></html></p>
```

And if you provide this input:

```html
<!DOCTYPE html>

<html><body>
some text here
</body></html>
```

You get this output:

```html
<!DOCTYPE html>

<p><html><body>
some text here
</body></html></p>
```

Notice that with no blank lines, the parser sees the entire block as a single paragraph. This is the confirmation that `html` and `body` tags are not considered block tags. And this is correct behavior as Markdown should only ever be content between `body` tags. To include a `body` (or its `html` parent) would generate invalid HTML. This is the Garbage-in, garbage-out principle. You can't expect valid output if you give it invalid input.

Note that the reference implementation (markdown.pl) follows the same behavior with one exception. It does not consider the doctype to be a block level element and also wraps it in a `p` element. If there is any bug here, it is that we recognize the doctype as a block level element. But personally, I'm fine leaving that as-is.

Note that this is a Markdown implementation which follows [these rules][1]. It is NOT a CommonMark implementation, nor will it ever be (CommonMark is not Markdown).

[1]: http://daringfireball.net/projects/markdown/syntax

**Comment by c-f-h** [comment]
The document you cited states this:

> Note that Markdown formatting syntax is not processed within block-level HTML tags. E.g., you can’t use Markdown-style *emphasis* inside an HTML block.

So the observed behavior is still a bug because the inner text within the `body` tag is treated as Markdown, which it shouldn't be according to the spec you cited.

**Comment by c-f-h** [comment]
Another, more relevant example: the input

```
<div>                                                                                                               
                                                                                                                    
some **text** here                                                                                                  
                                                                                                                    
</div>                                                                                                              
```
turns into

```
<div>                                                                                                               
                                                                                                                    
some **text** here                                                                                                  
                                                                                                                    
</div>
```

Clearly, `div` is a block tag, so the inner text should not be parsed as Markdown.

**Comment by waylan** [comment]
That depends on what is considered a "block-level HTML tag". As the reference implementation does not consider `body` or `html` to be block level, we don't either. And we will not change until the reference implementation does and/or the rules are updated.
